diff --git a/dist/types/icns.js b/dist/types/icns.js
index f2bfafef3723cb423b110304815e56e3f97f81e0..452d99f811bb333c42e8335c509215a8bfa151f7 100644
--- a/dist/types/icns.js
+++ b/dist/types/icns.js
@@ -82,6 +82,8 @@ exports.ICNS = {
         let imageOffset = SIZE_HEADER;
         let imageHeader = readImageHeader(input, imageOffset);
         let imageSize = getImageSize(imageHeader[0]);
+        if (imageHeader[1] < SIZE_HEADER)
+            throw new TypeError('Invalid ICNS entry length');
         imageOffset += imageHeader[1];
         if (imageOffset === fileLength)
             return imageSize;
@@ -93,6 +95,8 @@ exports.ICNS = {
         while (imageOffset < fileLength && imageOffset < inputLength) {
             imageHeader = readImageHeader(input, imageOffset);
             imageSize = getImageSize(imageHeader[0]);
+            if (imageHeader[1] < SIZE_HEADER)
+                throw new TypeError('Invalid ICNS entry length');
             imageOffset += imageHeader[1];
             result.images.push(imageSize);
         }
